### PR TITLE
run pdb check only if replicas >= 3

### PR DESCRIPTION
**Pull request by monotek**
Signed-off-by: André Bauer <andre.bauer@staffbase.com>

## Description

* run pdb check only if replicas >= 3 as pdb are usualy not defined for single pod deployments
* also add StatefulSet as resource
* use any in rule

## Checklist

```
kyverno test . 

Executing require-pdb...
applying 1 policy to 2 resources... 

│───│─────────────│─────────────│──────────────────────────────────────│────────│
│ # │ POLICY      │ RULE        │ RESOURCE                             │ RESULT │
│───│─────────────│─────────────│──────────────────────────────────────│────────│
│ 1 │ require-pdb │ require-pdb │ default/Deployment/nginx-deploy-pass │ Pass   │
│ 2 │ require-pdb │ require-pdb │ default/Deployment/nginx-deploy-skip │ Pass   │
│───│─────────────│─────────────│──────────────────────────────────────│────────│

Test Summary: 2 tests passed and 0 tests failed


```

- [] I have read the [policy contribution guidelines](https://github.com/kyverno/policies/blob/main/README.md#contribution).
- [] I have added test manifests and resources covering both positive and negative tests that prove this policy works as intended.

### CODE REVIEW

File: other/require_pdb/require_pdb.yaml [hunk 1]
```diff
@@ -23,10 +23,13 @@ spec:
...
         - Deployment
     preconditions:
-      any:
+      all:
       - key: "{{request.operation || 'BACKGROUND'}}"
         operator: Equals
         value: CREATE
+      - key: "{{ request.object.spec.replicas }}"
+        operator: GreaterThanOrEquals
+        value: "3"
```
Comment (chipzoller) [inline other/require_pdb/require_pdb.yaml:32]: You want to leave this unquoted so it's an integer.

**Review by chipzoller — CHANGES_REQUESTED** [review]
Minor thing. And as an "extra", it would be great if you could add maybe a couple more tests, including a negative test.

**Comment by monotek** [comment]
update the tests so one example with "skip" is available too.

File: other/require_pdb/require_pdb.yaml [hunk 2]
```diff
@@ -17,26 +17,29 @@ spec:
...
-      message: "There is no corresponding PodDisruptionBudget found for this Deployment."
-      deny:
-        conditions:
-          any:
-          - key: "{{pdb_count}}"
-            operator: LessThan
-            value: 1
\ No newline at end of file
+    - name: require-pdb
+      match:
```
Comment (chipzoller) [inline other/require_pdb/require_pdb.yaml:21]: Since you're updating this, go ahead and add `any`/`all` under the `match` object here to bring it current.

File: other/require_pdb/require_pdb.yaml [hunk 3]
```diff
@@ -17,26 +17,29 @@ spec:
...
+          kinds:
+            - Deployment
+      preconditions:
+        all:
+          - key: "{{request.operation || 'BACKGROUND'}}"
+            operator: Equals
+            value: CREATE
+          - key: "{{ request.object.spec.replicas }}"
+            operator: GreaterThanOrEquals
+            value: "3"
```
Comment (chipzoller) [inline other/require_pdb/require_pdb.yaml:32]: Don't need quotes around integers.

**Review by chipzoller — CHANGES_REQUESTED** [review]

File: other/require_pdb/require_pdb.yaml (hunk 2 above)
Comment (monotek) [reply other/require_pdb/require_pdb.yaml:21]: done

File: other/require_pdb/require_pdb.yaml (hunk 3 above)
Comment (monotek) [reply other/require_pdb/require_pdb.yaml]: done

**Review by chipzoller — APPROVED** [review]
